### PR TITLE
[JIT] Trace module calls

### PR DESCRIPTION
**Pull request by jamesr66a**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#28474 [JIT] Trace module calls**
* #28864 [JIT] API for finding a common ancestor block for a pair of nodes

Example:

```
import torch

torch._C._jit_set_inline_everything_mode(False)

class Bar(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.relu = torch.nn.ReLU()

    def forward(self, x):
        return self.relu(torch.neg(self.relu(x)))

class Foo(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.bar = Bar()
        self.linear = torch.nn.Linear(3, 4)

    def forward(self, x):
        return self.linear(self.bar(x))

foo = Foo()
x = torch.rand(3, 3)
traced = torch.jit.trace(foo, (x,))

with open('out.txt', 'w') as f:
    f.write(traced._c.dump_to_str())

scripted = torch.jit.script(foo)
with open('out_script.txt', 'w') as f:
    f.write(scripted._c.dump_to_str())

```

Trace result: https://gist.github.com/jamesr66a/7985c7f2b4a956b89f9de53b8e1f4d60
Script result: https://gist.github.com/jamesr66a/60bb962682871dc3dda855dc5c0528f3

TODOs:
* Fix tests
* Make stack trace actually print something

Differential Revision: [D18273353](https://our.internmc.facebook.com/intern/diff/D18273353)

### CODE REVIEW

File: torch/csrc/jit/passes/fixup_trace_scope_blocks.cpp [hunk 1]
```diff
@@ -0,0 +1,350 @@
...
+}
+
+void FixupTraceScopeBlocks(
+    std::shared_ptr<Graph>& graph,
+    script::Module* self) {
+  // Gather level information about blocks in the graph.
+  // Assign each block a number, with graph->block() gettng
+  // 0 and each sub-block gets a higher number based on how
+  // deeply nested it is.
+  BlockInfoMap block_info;
```
Comment (zdevito) [inline torch/csrc/jit/passes/fixup_trace_scope_blocks.cpp]: We should put level information into the Block itself. It does not change once a block is created so it is just `owningNode()->block()->level() + 1`

File: torch/csrc/jit/passes/fixup_trace_scope_blocks.cpp [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+  Value* v;
+  int level;
+};
+using RemappingTable = std::unordered_map<Value*, RemappedValueInfo>;
+
+void makeDefsDominateUses(
+    Block* b,
+    RemappingTable& remap,
+    const BlockInfoMap& block_info) {
+  auto find_common_ancestor_block_level = [&](Node* n) {
```
Comment (zdevito) [inline torch/csrc/jit/passes/fixup_trace_scope_blocks.cpp]: I think this should also be an API on Node, and can be more easily implemented when the level of each node is known pseudo code:

`while (one level is deeper than another) { go up one level }; while (block_lhs != block_rhs) { make both nodes go up a level; }

File: torch/csrc/jit/passes/fixup_trace_scope_blocks.cpp [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+
+struct RemappedValueInfo {
+  Value* v;
+  int level;
+};
+using RemappingTable = std::unordered_map<Value*, RemappedValueInfo>;
+
+void makeDefsDominateUses(
+    Block* b,
+    RemappingTable& remap,
```
Comment (zdevito) [inline torch/csrc/jit/passes/fixup_trace_scope_blocks.cpp]: Needs a comment about what this holds.

File: torch/csrc/jit/passes/fixup_trace_scope_blocks.cpp [hunk 4]
```diff
@@ -0,0 +1,350 @@
...
+
+  auto process_node = [&](Node* n) {
+    // inspect uses
+    for (size_t i = 0; i < n->inputs().size(); ++i) {
+      Value* inp = n->inputs()[i];
+      if (inp->node()->owningBlock() != b) {
+        int ancestor_level = find_common_ancestor_block_level(inp->node());
+
+        Block* origin_block = inp->node()->owningBlock();
+        if (ancestor_level == block_info.at(origin_block)) {
```
Comment (zdevito) [inline torch/csrc/jit/passes/fixup_trace_scope_blocks.cpp]: I don't think this special case needs to exist. It seems like it is just the degenerate version of the code below.

File: torch/csrc/jit/passes/fixup_trace_scope_blocks.cpp [hunk 5]
```diff
@@ -0,0 +1,350 @@
...
+          remap[inp] = {inp, block_info.at(origin_block)};
+        } else {
+          Value* v_itr = inp;
+          Block* b = origin_block;
+
+          // Start off where we left off if there's an existing mapping
+          if (remap.count(inp)) {
```
Comment (zdevito) [inline torch/csrc/jit/passes/fixup_trace_scope_blocks.cpp]: Because I don't know what remap holds (e.g. what does the level mean), I am finding it hard to follow this logic.

File: torch/csrc/jit/passes/fixup_trace_scope_blocks.cpp [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+  }
+}
+
+struct RemappedValueInfo {
+  Value* v;
+  int level;
+};
+using RemappingTable = std::unordered_map<Value*, RemappedValueInfo>;
+
+void makeDefsDominateUses(
```
Comment (zdevito) [inline torch/csrc/jit/passes/fixup_trace_scope_blocks.cpp]: This transform might be easier to follow as a class because it already contains lambda functions, and there is a several shared members (remap, block_info).

File: torch/csrc/jit/passes/fixup_trace_scope_blocks.cpp [hunk 7]
```diff
@@ -0,0 +1,350 @@
...
+
+        // Make node outputs a single tuple;
+        std::vector<TypePtr> types;
+        for (size_t i = 0; i < n->outputs().size(); ++i) {
+          types.push_back(n->output(0)->type());
+        }
+        Value* tup_output = n->addOutput()->setType(TupleType::create(types));
+        Node* tup_unpack = g->createTupleUnpack(tup_output)->insertAfter(n);
+        for (size_t i = 0; i < tup_unpack->outputs().size(); ++i) {
+          n->output(0)->replaceAllUsesWith(tup_unpack->output(i));
```
Comment (zdevito) [inline torch/csrc/jit/passes/fixup_trace_scope_blocks.cpp]: This is n^2 to erase the first output each time.

File: torch/csrc/jit/passes/fixup_trace_scope_blocks.cpp [hunk 8]
```diff
@@ -0,0 +1,350 @@
...
+        }
+        return remaps[v];
+      });
+      LintGraph(graph);
+      n->g_(attr::Subgraph, graph);
+      n->eraseBlock(0);
+    }
+  }
+
+  // Inline base __module call
```
Comment (zdevito) [inline torch/csrc/jit/passes/fixup_trace_scope_blocks.cpp]: I don't understand what is going on here. Seems like something that should be handled separately. Comment explains what but not why

File: torch/csrc/jit/passes/fixup_trace_scope_blocks.cpp [hunk 9]
```diff
@@ -0,0 +1,350 @@
...
+    Node* n = *n_itr++;
+    if (n->kind() == prim::FakeScopeBlock && n->s(attr::scope) == "__module") {
+      inlineCallTo(n, *n->g(attr::Subgraph));
+    }
+  }
+}
+
+namespace {
+
+std::vector<std::string> splitModuleQualname(const std::string& s) {
```
Comment (zdevito) [inline torch/csrc/jit/passes/fixup_trace_scope_blocks.cpp]: QualifiedName already implements this.

File: torch/csrc/jit/passes/fixup_trace_scope_blocks.cpp [hunk 10]
```diff
@@ -0,0 +1,350 @@
...
+          callee_val = g->insertGetAttr(callee_val, sub_atoms[i]);
+          callee_mod = callee_mod.get_module(sub_atoms[i]);
+        }
+      }
+
+      createMethodCalls(n->g(attr::Subgraph), callee_mod, sub_atoms);
+
+      Function* f = nullptr;
+
+      // Find a unique name to add this method as
```
Comment (zdevito) [inline torch/csrc/jit/passes/fixup_trace_scope_blocks.cpp]: Please make naming mangling into a function.

File: torch/csrc/jit/passes/fixup_trace_scope_blocks.cpp [hunk 11]
```diff
@@ -0,0 +1,350 @@
...
+
+} // namespace
+
+void createMethodCalls(
+    const std::shared_ptr<Graph>& g,
+    script::Module this_level_self,
+    std::vector<std::string> prefix) {
+  Value* self;
+  if (g->inputs().size() == 0 ||
+      g->inputs()[0]->type() != this_level_self.type()) {
```
Comment (zdevito) [inline torch/csrc/jit/passes/fixup_trace_scope_blocks.cpp]: I don't know what this is suppose to be checking for, but pointer comparison of the types is most likely wrong.

Comment (zdevito) [reply torch/csrc/jit/passes/fixup_trace_scope_blocks.cpp]: Something feels wrong here. There is no guarantee that something with the same type as self is the self module (we should not assume they are singletons).  I feel like Module self should just be the first argument everywhere so it doesn't have to be recovered here.

File: torch/csrc/jit/passes/fixup_trace_scope_blocks.cpp [hunk 12]
```diff
@@ -0,0 +1,350 @@
...
+    // Lambda lift Values (i.e. add Graph inputs for the purpose of
+    // referencing values that dominate the block) and convert
+    // the block to a Graph. blocks()[0] on each FakeScopeBlock then
+    // appears as a Graph attribute attr::Subgraph
+    lambdaLiftBlocksAndConvertToGraph(graph->block());
+    // Register the attr::Subgraph Graph values as Functions in the
+    // class compilation unit and register that Function as a method
+    // on the corresponding Module in the Module hierarchy. Note that we
+    // unique the methods by naming them forward, forward1, forward2...
+    createMethodCalls(graph, *self, {"__module"});
```
Comment (zdevito) [inline torch/csrc/jit/passes/fixup_trace_scope_blocks.cpp]: Can you make an argument for the algorithmic complexity of this whole process? I want to make sure it is not N^2 as the module depth grows, but I cannot tell by inspection. Most of these passes are recursive so it is not clear what happens when we repeatedly apply them.

File: torch/jit/__init__.py [hunk 13]
```diff
@@ -280,7 +269,10 @@ def get_trace_graph(f, args=(), kwargs=None, _force_outplace=False,
         kwargs = {}
     if not isinstance(args, tuple):
         args = (args,)
-    return ONNXTracedModule(f, _force_outplace, return_inputs, _return_inputs_states)(*args, **kwargs)
+    torch.jit._trace_module_map = {}
```
Comment (zdevito) [inline torch/jit/__init__.py]: You get _one_ place where this global is set/reset.  It is really hard to know where it has to be manipulated at the moment because it exists in three different places. Pass it in as an argument up until it is no longer feasibly to do so.

File: torch/csrc/jit/passes/fixup_trace_scope_blocks.cpp [hunk 14]
```diff
@@ -0,0 +1,350 @@
...
+  for (Node* n : b->nodes()) {
+    if (n->kind() == prim::FakeScopeBlock) {
+      inspectBlocks(n->blocks()[0], block_info, level + 1);
+    }
+  }
+}
+
+struct RemappedValueInfo {
+  Value* v;
+  int level;
```
Comment (zdevito) [inline torch/csrc/jit/passes/fixup_trace_scope_blocks.cpp]: not sure why this needs to be tracked.

**Review by zdevito — COMMENTED** [review]
This is looking pretty good. My overall comment is that is pretty hard to follow the logic in the pass due to some missing comments. Putting the find common node logic in ir.cpp would help, as would making makiDefsDominateUses into a  class.

File: torch/nn/modules/module.py [hunk 15]
```diff
@@ -500,32 +501,30 @@ def register_forward_hook(self, hook):
...
-        else:
-            tracing_state.push_scope(self._get_name())
-        tracing_state._traced_module_stack.append(self)
+        recording_scopes = torch.jit._trace_module_map is not None
+        if recording_scopes:
+            name = torch.jit._trace_module_map[self] if self in torch.jit._trace_module_map else None
+            if name:
+                cur_scope_name = tracing_state.curr_scope()
+                if not name.startswith(cur_scope_name):
+                    warnings.warn('Called module {} from {}, but these modules are not in a '
```
Comment (zdevito) [inline torch/nn/modules/module.py]: This restriction isn't necessary. The nesting of ScopeBlocks does not have to match the nesting of the module hierarchy. In the cases where it doesn't the module whose method _is_ being called is going to need to be passed in as a non-self argument to calling method. But this is something that is already handled by the capture process. Essentially this can be done with two steps:

1. Before converting subblocks, just add the correct `self` argument to the block, which is determined by the _trace_module_map.
2. Also before converting subblocks, replace all uses of variables inside the immediate subblock of values that are actually attributes of self for that subblock with their getattr equivalents. This would make sure we do not pass self.weight as an argument to the method but look it up from the module. This will also make all those `self` arguments we added in step 1 into get attrs on the parent when possible.

In cases where we do not have a hierarchy, then modules will get passed around as arguments that are not self, but the code will otherwise be correct.

**Review by zdevito — COMMENTED** [review]
Haven't finished reading the actual transform code, but see the one comment so far below.

**Comment by jamesr66a** [comment]
Closing because ghstack fuckery
